core/validation/specific_validations/towns_fund_round_four.py

```diff
@@ -268,6 +270,86 @@ def validate_leading_factor_of_delay(
         ]
 
 
+def validate_funding_spent(workbook: dict[str, pd.DataFrame]) -> list["TownsFundRoundFourValidationFailure"] | None:
+    """Validates that total reported funding spent on the sheet is less than the amount specified as allocated in a
+    separate sheet stored on S3.
+
+    For Town deal funds this is done on a per-project basis whereas for future High Street Funds this is done
+    per-programme leading to slightly different logics for both.
+
+    :param workbook: A dictionary where keys are sheet names and values are pandas
+                     DataFrames representing each sheet in the Round 4 submission.
+    :return: ValidationErrors
+    """
+    # Pull programme and project indexes from the workbook
+    programme_id = workbook["Programme_Ref"]["Programme ID"][0]
+    fund_type = programme_id[:2]
+    project_ids = workbook["Project Details"]["Project ID"]
+    funding_df = workbook["Funding"]
+
+    # pull funding spent for individual projects and programme wide total
+    funding_spent = {project: spend_per_project(funding_df, project) for project in project_ids}
+    funding_spent[programme_id] = sum(funding_spent.values())
+
+    allocated_file = s3.get_file("data-store-file-assets-dev", f"{fund_type}-grant-awarded.csv")
+    funding_allocated = pd.read_csv(allocated_file, index_col="Index Code")["Grant Awarded"]
+
+    funding_spent_failures = []
+    if fund_type == "HS":
+        # check funding against programme wide funding allocated for Future High Street Fund submissions
+        ids_to_check = [programme_id]
+        for idx in ids_to_check:
+            if funding_spent[idx] > funding_allocated[idx]:
+                funding_spent_failures.append(
+                    TownsFundRoundFourValidationFailure(
+                        tab="Funding Profiles",
+                        section="Project Funding Profiles",
+                        message=(
+                            f"[Template message] You spent {funding_spent[idx]} "
+                            f"but were only allocated {funding_allocated[idx]}"
+                        ),
+                    )
+                )
+    else:
+        # check funding against individual project funding allocated for Towns Deal submissions
+        ids_to_check = project_ids
+        for idx in ids_to_check:
+            if funding_spent[idx] > funding_allocated[idx]:
+                funding_spent_failures.append(
+                    TownsFundRoundFourValidationFailure(
+                        tab="Funding Profiles",
+                        section=f"Project Funding Profiles - Project {get_project_number(idx)}",
+                        message=(
+                            f"[Template message] You spent {funding_spent[idx]} "
+                            f"but were only allocated {funding_allocated[idx]}"
+                        ),
+                    )
+                )
```

Comment:
There's a fair bit of duplication here. 

Can we set `ids_to_check` depending on `fund_type` and then reuse the same code to check and capture any failures?

I suppose section is depended on `fund_type` as well but we could also add that as an inline if else?